### DISCUSSION TITLE
Optimal strategy for implementing splash/login/etc. forms in large-scale enterprise apps

### DISCUSSION THREAD

**Discussion by JDr111** [AvaloniaUI/Avalonia#14870] [General]
Hello everyone,
I'm fairly new to Avalonia and desktop MVVM in general, and I need to clarify/confirm whether I'm thinking correctly about this topic.

Suppose we're starting developing an enterprise app with a large emphasis on user comfort, stability, maintainability and security (the last one often contradicts the first one, but alas, we have to live with it). The app must authenticate on the backend server using Bearer tokens obtained via OAuth/OpenID Connect protocols.

In this particular project, there's a strict requirement to implement a sort of "introductory flow" the user goes through when the app starts. This flow goes as follows:

1. Splash view - load configuration, resources, do whatever else that needs to be done at this stage. There's no interactivity with the user (spare for maybe a "cancel" button), and the flow progresses to the next stage once all loading is done (that is, the view automatically changes to the next one).
2. Login view - An interactive view containing a button, which opens an external browser, prompting the user to sign in to an external authentication server. Upon success, an access token is delivered to the app via a loopback redirection (see [RFC 8252](https://datatracker.ietf.org/doc/html/rfc8252) for more info). Once the app obtains an access token, the flow automatically progresses to the next stage.
3. Project selection view - The user is prompted to select a project to work with. Once the user does so, the project selection screen closes and the main window appears, ending the "introductory flow".

I have each of the three "views" implemented and functional separately in "lab" mini-projects. The main question here is - what is a good way to connect the three views into one flow as described?

Essentially, it boils down to implementing some sort of a routing/navigation mechanism that would allow for switching between the views. Something similar to [one of the samples](https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample), but with the key difference that the view switching is triggered by code-behind (e.g. an event raised by the hosted views (splash, login, project selection), rather than the user clicking a button (with the exception of the project selection view)).

I had two thoughts:

- Taking inspiration from the sample, I could implement a navigation view/control whose content binds to a property in its ViewModel (probably discarding the view locator, as it's not really needed in this case?). This navigation view would thus host the splash, login and project selection views, allowing for fancy transitions using the TransitioningContentControl. However, with this approach, each of the three views would need a way to send a signal to the navigation control to change the view to another one. This isn't that hard, this can be done with a routed event that's defined in the navigation view. This means I'd have to handle all logic of each view in its ViewModel, and then have the view automatically trigger the view switch once done, using the navigation's routed event. That also means the hosted views would "know" about the navigation view, since they'd raise its routed event (even though it's a static field, technically), and possibly even create new instances of the new views. Once the flow is finished, the window hosting the navigation view would close, and the main window would appear. Changing the app's "MainWindow" property happens in `App.xaml.cs`.

- The second thought is to implement each view in the "introductory flow" in a separate window. This would sacrifice having fancy transitions between the views, but the views would no longer need to know about each other and would therefore be completely separate entities, and all logic and window switching happens in `App.axaml.cs`. This is also probably easier to implement.

Are these good strategies to implement the "introductory flow" as specified? Which one is better, or how could they be improved? Is my way of thinking about this correct at all?

To better clarify the thought process, see the mind map:
![image](https://github.com/AvaloniaUI/Avalonia/assets/156440581/cbce347f-9abd-4128-b6a7-7f994f8e9e0e)

Any thoughts and ideas for improvements would be helpful. Thanks!

**Comment by thevortexcloud** [comment]
> . That also means the hosted views would "know" about the navigation view

There is no reason to do it this way. You can work entirely with view model/events. For example, take a look at ReactiveUIs router which uses ViewModel navigation.

https://docs.avaloniaui.net/docs/concepts/reactiveui/routing

The custom control provided by it simply just listens to change events on the routers. The view models have no idea the view even exists. As a result, you can simply just make each view model load the next one once it is in the completed state, the view will then see that change and will update the which template is loaded to display the newly selected view model. 

Remember: In MVVM the view is allowed to know about the view model, the viewmodel is not allowed to know about the view. As such your view is allowed to poke/watch your view model as much as it wants (although typically it should be avoided when possible since it can make porting a pain).

There are plenty of other ways to handle routing too. You can even make your own fairly easily. I made one awhile back that worked using URL like paths so I could just put a command parameter into a button to change which view was made active.

Anyway, it really sounds like you also want some sort of state machine to manage which of your views should be active at any time. Typically you would make that a dedicated class that is only loosely coupled to your views and/or view models. If you use a router it just needs to know about your VMs and expose some sort of state change event that can be used to update your router.

My advice is if you can and it makes sense for your app, stick to a SPA style app. It makes porting to other platforms significantly easier.

**Reply by JDr111** [reply]
Thanks for the insights.

I forgot to mention, I'm using the Community MVVM Toolkit for this project, not ReactiveUI, as it seems to be a bit more intuitive to use, especially for someone who's still getting to know things. But I assume the same thing can be done with both.

Seems like I basically want to implement a routing mechanism that resolves ViewModels into Views, and have the "hosted views" raise an event (via whatever mechanism) when the view switch should happen, since the views need to be able to switch without user input.

Looking at ReactiveUI's router, the key takeaways are the following:

- There's one control that monitors the currently selected view model, and displays the corresponding view. This can be done by binding the Content property to a ViewModel, and have the view locator resolve it into a View.
- Each view model needs to be "routable". Looking at how ReactiveUI does it, each routable ViewModel has a reference to IScreen, which according to the docs should be the router's ViewModel, or the main window. I assume this is how the routable ViewModels can notify the router to switch the views. Whether that notification is made automatically or upon a button click depends on the particular ViewModel.
- I agree that views should minimize directly messing with the ViewModels, ideally limiting it only to bindings. Depends on the particular situation, this won't always be possible.

That clears up a lot of things. I'll see what I can come up with now. Thanks again!
